pkg/remote/codec/unknown/unknown.go

```diff
@@ -0,0 +1,248 @@
+package unknown
+
+import (
+	"context"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	thrif "github.com/apache/thrift/lib/go/thrift"
+	"github.com/cloudwego/dynamicgo/thrift"
+	"github.com/cloudwego/kitex/pkg/protocol/bthrift"
+	"github.com/cloudwego/kitex/pkg/remote"
+	"github.com/cloudwego/kitex/pkg/remote/codec"
+	"github.com/cloudwego/kitex/pkg/remote/codec/perrors"
+	unknowns "github.com/cloudwego/kitex/pkg/remote/codec/unknown/service"
+	"github.com/cloudwego/kitex/pkg/rpcinfo"
+	"github.com/cloudwego/kitex/pkg/serviceinfo"
+)
+
+// UnknownCodec implements PayloadCodec
+type unknownCodec struct {
+	Codec remote.PayloadCodec
+}
+
+func NewUnknownCodec(code remote.PayloadCodec) remote.PayloadCodec {
+	return &unknownCodec{code}
+}
+
+func (c unknownCodec) Marshal(ctx context.Context, msg remote.Message, out remote.ByteBuffer) error {
+	nw, _ := out.(remote.NocopyWrite)
+	ink := msg.RPCInfo().Invocation()
+	data := msg.Data()
+	msgType := msg.MessageType()
+	if data, ok := data.(*unknowns.Result); ok {
+		//svcInfo := message.ServiceInfo()
+		//unknowns.SetServiceInfo(svcInfo, message.ProtocolInfo().CodecType, data.Method, data.Method)
+		if len(data.Success) == 0 {
+			return errors.New("unknown messages cannot be empty")
+		}
+		if msg.MessageType() == remote.Exception {
+			return c.Codec.Marshal(ctx, msg, out)
+		}
+		if ink, ok := ink.(rpcinfo.InvocationSetter); ok {
+			ink.SetMethodName(data.Method)
+			ink.SetServiceName(data.ServiceName)
+		} else {
+			return errors.New("the interface Invocation doesn't implement InvocationSetter")
+		}
+		if msg.ProtocolInfo().CodecType == serviceinfo.Thrift {
```

Comment:
Please refactor the two branches to be `marshalThrift` and `marshalProtobuf`. And as they are almost the same except for the magic, I guess only one method is enough?

Reply:
Avoid long functions to make your code easier to read.